### PR TITLE
[TRYC-49] 게시글 상태 변경 기능 개발 & [TRYC-151] 동네 거리 계산 로직 구현

### PR DESCRIPTION
**Pull request by ybkim-dev**
### 변경 내용 요약
- 지난 회의 시간에 정의한대로 주변 동네 거리 계산 로직을 구현
- 게시글 상태 변경 기능 개발

### 작업한 내용
- 지난 회의 시간에 정의한대로 주변 동네 거리 계산 로직을 구현했습니다.
  - 모든 동네에 대해서 단계 별로 가까운 동네들을 리스트업하고 id와 name을 mongoDB에 저장했습니다.
  - MemberTown 도메인을 통해 선택한 동네 기준 거리 레벨 별로 주변 동네의 id 리스트를 가져올 수 있도록 기능 구현했습니다.
  - 현재 저장 및 조회 성공 테스트만 작성되어 있습니다. 2차 스프린트 이내에 개발해야할 내용이 조금 남아서 실패 상황 테스트는 3차 스프린트에 구현할 예정입니다.😓
- 게시글 판매 상태 변경 기능을 개발했습니다.
  - 판매 상태에 대해서만 입력할 수 있도록 예외처리 해두었습니다.
  - 현재 테스트 코드 작성 완료 상태입니다.

### 관련 링크
- [TRYC-151](https://cloudwi.atlassian.net/browse/TRYC-151)
- [TRYC-49](https://cloudwi.atlassian.net/browse/TRYC-49)

### 기타 사항
기타 공유하고 싶은 점을 적어주세요.

[TRYC-151]: https://cloudwi.atlassian.net/browse/TRYC-151?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[TRYC-49]: https://cloudwi.atlassian.net/browse/TRYC-49?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by cloudwi — APPROVED** [review]

File: src/main/java/com/dangdang/server/domain/post/application/PostService.java [hunk 1]
```diff
@@ -56,23 +60,36 @@ public PostsSliceResponse findPostsForSlice(PostSliceRequest postSliceRequest,
...
+    List<String> imageUrls = postImageService.findPostImagesByPostId(postId);
+    return PostDetailResponse.from(foundPost, foundPost.getMember(), imageUrls);
+  }
+
+  @Transactional
+  public PostDetailResponse updatePostStatus(Long postId, PostUpdateStatusRequest postUpdateStatusRequest, Long authorId) {
+    Post post = postRepository.findPostDetailById(postId)
+        .orElseThrow(() -> new PostNotFoundException(POST_NOT_FOUND));
+
+    if(!Objects.equals(post.getMemberId(), authorId)) {
```
Comment (Albatross3) [inline src/main/java/com/dangdang/server/domain/post/application/PostService.java:87]: public boolean equals(Object obj) {
        if (obj instanceof Long) {
            return value == ((Long)obj).longValue();
        }
        return false;
    }
Long class 에도 equals가 있어서 그것을 사용해도 될 것 같습니다~

File: src/main/java/com/dangdang/server/domain/post/dto/request/PostUpdateStatusRequest.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+  public PostUpdateStatusRequest {
+    verifyValidStatus(status);
+  }
+
+  private void verifyValidStatus(StatusType status) {
+    switch(status) {
+      case SELLING, RESERVED, COMPLETED -> {}
+      default -> throw new InvalidStatusTypeException(INVALID_POST_STATUS);
+    }
```
Comment (Albatross3) [inline src/main/java/com/dangdang/server/domain/post/dto/request/PostUpdateStatusRequest.java:20]: StatusType enum 클래스에서 이미 valueOf 메서드를 통해 enum 값에 해당하는지 검증하기 때문에 검증을 또 안해도 되지 않을까요?

File: src/main/java/com/dangdang/server/domain/postImage/application/PostImageService.java [hunk 3]
```diff
@@ -22,10 +22,13 @@ public PostImageService(PostImageRepository postImageRepository) {
...
 
   @Transactional(propagation = Propagation.MANDATORY)
-  public void savePostImage(Post post, PostImageRequest postImageRequest) {
+  public List<String> savePostImage(Post post, PostImageRequest postImageRequest) {
     List<String> urls = postImageRequest.getUrl();
     urls.stream().map(url -> PostImageRequest.toPostImage(post, url))
         .forEach(postImageRepository::save);
+    return postImageRepository.findPostImagesByPostId(post.getId()).stream()
+        .map(PostImage::getUrl)
+        .collect(Collectors.toList());
```
Comment (Albatross3) [inline src/main/java/com/dangdang/server/domain/postImage/application/PostImageService.java:31]: 위에 있는 urls 변수를 return 하면 안되나요?
postImage를 저장할 때마다 조회하는 쿼리를 추가적으로 날려야하는 것을 한 번 생각해봐야할 것 같습니다!

File: src/test/java/com/dangdang/server/domain/post/application/PostServiceTest.java [hunk 4]
```diff
@@ -58,17 +58,19 @@ void postSaveTest() {
...
 
-    PostDetailResponse postDetailResponse = postService.findPostDetailById(
-        savedPostResponse.getId());
-    assertThat(savedPostResponse).usingRecursiveComparison()
-        .isEqualTo(postDetailResponse.getPostResponse());
+    PostDetailResponse foundPost = postService.findPostDetailById(savedPostResponse.getPostId());
+    assertThat(savedPostResponse.getPostResponse()).usingRecursiveComparison()
+        .isEqualTo(foundPost.getPostResponse());
+    assertThat(savedPostResponse.getMember()).usingRecursiveComparison()
+        .isEqualTo(foundPost.getMember());
```
Comment (Albatross3) [inline src/test/java/com/dangdang/server/domain/post/application/PostServiceTest.java:73]: url도 검증해주시죠!

File: src/test/java/com/dangdang/server/domain/post/application/PostServiceTest.java [hunk 5]
```diff
@@ -89,18 +90,18 @@ void findPostDetailById() {
...
+    PostDetailResponse savedPostResponse = postService.savePost(postSaveRequest, loginMember);
 
-    PostDetailResponse foundPost = postService.findPostDetailById(savedPostResponse.getId());
+    PostDetailResponse foundPost = postService.findPostDetailById(savedPostResponse.getPostId());
     assertThat(foundPost).isNotNull();
     assertThat(foundPost.getImageUrls()).hasSize(2);
-    assertThat(foundPost.getImageUrls()).usingRecursiveComparison();
   }
 
   @Test
```
Comment (Albatross3) [inline src/test/java/com/dangdang/server/domain/post/application/PostServiceTest.java:107]: 이 테스트는 위와 "게시글을 상세 조회 할 수 있다." 테스트에서 충분히 가능할 것 같습니다!
로직면에서 중복적인 것 같아서 리뷰 남겼습니다~

**Review by Albatross3 — APPROVED** [review]
기웅님,영빈님 고생하셨습니다!
리뷰에 더불어 추가적으로 드리고 싶은 말은 PostService Test에서 edge case 에 대한 test 코드가 더 많이 작성되면 좋을 것 같습니다~

File: src/main/java/com/dangdang/server/domain/post/application/PostService.java (hunk 1 above)
Comment (ybkim-dev) [reply src/main/java/com/dangdang/server/domain/post/application/PostService.java:87]: 앗 그렇군요! 수정하겠습니다!!

File: src/main/java/com/dangdang/server/domain/post/dto/request/PostUpdateStatusRequest.java (hunk 2 above)
Comment (ybkim-dev) [reply src/main/java/com/dangdang/server/domain/post/dto/request/PostUpdateStatusRequest.java:20]: Statustype enum 클래스는 Enum 클래스에 해당되는 값이 들어왔는지를 확인합니다만 여기서는 SELLING, RESERVED, COMPLETED 이외의 값이 드어오면 에러를 뱉도록 설정했습니다.

File: src/main/java/com/dangdang/server/domain/postImage/application/PostImageService.java (hunk 3 above)
Comment (ybkim-dev) [reply src/main/java/com/dangdang/server/domain/postImage/application/PostImageService.java:31]: urls 에 가지고 있는 URL은 Permission denied 되어 읽을 수 없는 S3 이미지 링크입니다. 조회해주는 로직에서 읽을 수 있도록 permission을 주고 있어서 이렇게 작성했습니다!

File: src/test/java/com/dangdang/server/domain/post/application/PostServiceTest.java (hunk 4 above)
Comment (ybkim-dev) [reply src/test/java/com/dangdang/server/domain/post/application/PostServiceTest.java:73]: URL을 검증할 수 있는 방법이 없습니다..ㅠ URL이 실제 사진이 있는 파일있는 URL인지 검증하려했지만 그 부분은 실패했습니다.
URL 자체가 있는지에 대한 테스트는 PostRepository에서 진행해주고 있습니다.

**Review by ybkim-dev — COMMENTED** [review]
리뷰 감사합니다 동준님! ㅎㅎㅎ
